File: unidade1/derivacao.py
import math

# Fazendo uma função exemplo
def funcao_exemplo(x):
    return x**2 * math.log(x) + math.exp(x)

# Calculando as derivadas
# Derivadas Primeira
def derivada_primeira_forward(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'linear'):
        numerador = funcao(x + Dx) - funcao(x)
        denominador = Dx

    elif (erro == 'quadratico'):
        numerador = -funcao(x + 2*Dx) + 4 * funcao(x + Dx) - 3 * funcao(x)
        denominador = (2*Dx)

    elif (erro == 'cubico'):
        numerador = - 11 * funcao(x) + 18 * funcao(x + Dx) - 9 * funcao(x + 2*Dx) + 2 * funcao(x + 3*Dx)
        denominador = 6*Dx

    elif (erro == 'quarta_ordem'):
        numerador = - 25 * funcao(x) + 48 * funcao(x + Dx) - 36 * funcao(x + 2*Dx) + 16 * funcao(x + 3*Dx) - 3 * funcao(x + 4*Dx)
        denominador = 12*Dx

    return numerador / denominador

def derivada_primeira_backward(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'linear'):
        numerador = funcao(x) - funcao(x - Dx)
        denominador = Dx

    elif (erro == 'quadratico'):
        numerador = 3 * funcao(x) - 4 * funcao(x - Dx) + funcao(x - 2*Dx)
        denominador = 2*Dx

    elif (erro == 'cubico'):
        numerador = 11 * funcao(x) - 18 * funcao(x - Dx) + 9 * funcao(x - 2*Dx) - 2 * funcao(x - 3*Dx)
        denominador = 6*Dx

    elif (erro == 'quarta_ordem'):
        numerador = 25 * funcao(x) - 48 * funcao(x - Dx) + 36 * funcao(x - 2*Dx) - 16 * funcao(x - 3*Dx) + 3 * funcao(x - 4*Dx)
        denominador = 12*Dx 

    return numerador / denominador 

def derivada_primeira_central(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'quadratico'):
        numerador = funcao(x + Dx) - funcao(x - Dx)
        denominador = 2*Dx
    
    elif (erro == 'quarta_ordem'):
        numerador = - funcao(x + 2*Dx) + 8 * funcao(x + Dx) - 8 * funcao(x - Dx) + funcao(x - 2*Dx)
        denominador = 12*Dx

    return numerador / denominador

# Derivadas Segunda
def derivada_segunda_forward(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'linear'):
        numerador = funcao(x) - 2 * funcao(x + Dx) + funcao(x + 2*Dx)
        denominador = Dx*Dx

    elif (erro == 'quadratico'):
        numerador = 2 * funcao(x) - 5 * funcao(x + Dx) + 4 * funcao(x + 2*Dx) -  funcao(x + 3*Dx)
        denominador = Dx*Dx

    elif (erro == 'cubico'):
        numerador = 35 * funcao(x) - 104 * funcao(x + Dx) + 114 * funcao(x + 2*Dx) - 56 * funcao(x + 3*Dx) + 11 * funcao(x + 4*Dx)
        denominador = 12 * Dx * Dx

    return numerador / denominador

def derivada_segunda_backward(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'linear'):
        numerador = funcao(x) - 2 * funcao(x - Dx) + funcao(x - 2*Dx)
        denominador = Dx*Dx

    elif (erro == 'quadratico'):
        numerador = 2 * funcao(x) - 5 * funcao(x - Dx) + 4 * funcao(x - 2*Dx) - funcao(x - 3*Dx)
        denominador = Dx*Dx

    elif (erro == 'cubico'):
        numerador = 35 * funcao(x) - 104 * funcao(x - Dx) + 114 * funcao(x - 2*Dx) - 56 * funcao(x - 3*Dx) + 11 * funcao(x - 4*Dx)
        denominador = 12*Dx*Dx

    return numerador / denominador

def derivada_segunda_central(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'quadratico'):
        numerador = funcao(x + Dx) - 2 * funcao(x) + funcao(x - Dx)
        denominador = Dx*Dx

    elif (erro == 'quarta_ordem'):
        numerador = - funcao(x + 2*Dx) + 16 * funcao(x + Dx) - 30 * funcao(x) + 16 * funcao(x - Dx) - funcao(x - 2*Dx)
        denominador = 12*Dx*Dx

    return numerador / denominador

# Derivada Terceira
def derivada_terceira_forward(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'linear'):
        numerador = + funcao(x + 3*Dx) - 3 * funcao(x + 2*Dx) + 3 * funcao(x + Dx) - funcao(x)
        denominador = Dx**3

    elif (erro == 'quadratico'):
        numerador = - 5 * funcao(x) + 18 * funcao(x + Dx) - 24 * funcao(x + 2*Dx) + 14 * funcao(x + 3*Dx) - 3 * funcao(x + 4*Dx)
        denominador = 2*Dx**3

    return numerador / denominador

def derivada_terceira_backward(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'linear'):
        numerador = funcao(x) - 3 * funcao(x - Dx) + 3 * funcao(x - 2*Dx) - funcao(x - 3*Dx)
        denominador = Dx**3

    elif (erro == 'quadratico'):
        numerador = 5 * funcao(x) - 18 * funcao(x - Dx) + 24 * funcao(x - 2*Dx) - 14 * funcao(x - 3*Dx) + 3 * funcao(x - 4*Dx)
        denominador = 2*Dx**3

    return numerador / denominador

def derivada_terceira_central(erro, x, Dx, funcao = funcao_exemplo):
    if (erro == 'quadratico'):
        numerador = + funcao(x + 2*Dx) - 2 * funcao(x + Dx) + 2 * funcao(x - Dx) - funcao(x - 2*Dx)
        denominador = 2*Dx**3

    elif (erro == 'quarta_ordem'):
        numerador = - funcao(x + 3*Dx) + 8 * funcao(x + 2*Dx) - 13 * funcao(x + Dx) + 13 * funcao(x - Dx) - 8 * funcao(x - 2*Dx) + funcao(x - 3*Dx)
        denominador = 8*Dx**3

    return numerador / denominador

File: unidade1/java/Derivacao.java
package MetodosNumericosII.unidade1.java;

// Classe para realizar as operacoes de derivacao
public class Derivacao {

    // Definindo os atributos da classe
    private double x; 
    private double deltaX;
    private String funcao;

    // Construtor
    Derivacao(double x, double deltaX, String funcao){
        this.setX(x);
        this.setDeltaX(deltaX);
        this.setFuncao(funcao);
    }

    // Getters
    public double getX(){
        return x;
    }

    public double getDeltaX(){
        return deltaX;
    }

    public String getFuncao(){
        return funcao;
    }

    // Setters
    public void setX(double x){
        this.x = x;
    }

    public void setDeltaX(double deltaX){
        this.deltaX = deltaX;
    }

    public void setFuncao(String funcao){
        this.funcao = funcao;
    }

    // Criando os metodos para o calculo da derivada
        // Derivada primeira
    public double derivadaPrimeira(int tipo){

        // Declarando numerador e denominador
        // Deixo inicialmente null para nao dar erro de compilacao
        // Ja que um dos metodos sera escolhido, entao o numerador e o denominador sempre mudarao
        Double numerador = null;
        Double denominador = null;

        // Se for escolhido o tipo 1, entao sera calculado a derivada pelo metodo forward
        if (tipo == 1){
            numerador = ExecutarPython.converter_funcao(funcao, x + deltaX) 
            - ExecutarPython.converter_funcao(funcao, x);

            denominador = deltaX;
        }

        // Se for escolhido o tipo 2, entao sera calculado a derivada pelo metodo backward
        else if (tipo == 2){
            numerador = ExecutarPython.converter_funcao(funcao, x) 
            - ExecutarPython.converter_funcao(funcao, x - deltaX);

            denominador = deltaX;
        }

        // Se for escolhido o tipo 3, entao sera escolhido o metodo central
        else if (tipo == 3){
            numerador = ExecutarPython.converter_funcao(funcao, x + deltaX) 
            - ExecutarPython.converter_funcao(funcao, x - deltaX);

            denominador = 2 * deltaX;
        }

        // Calculando a reta secante
        double secante = numerador / denominador;
        return secante;

    }

        // Derivada segunda
    public Double derivadaSegunda(int tipo){

        // Declarando numerador e denominador
        Double numerador = null;
        Double denominador = null;

        // Se for escolhido o tipo 1, entao sera calculado a derivada pelo metodo forward
        if (tipo == 4){
            numerador = 
            ExecutarPython.converter_funcao(funcao, x + 2 * deltaX) 
            - 2 * ExecutarPython.converter_funcao(funcao, x + deltaX)
            + ExecutarPython.converter_funcao(funcao, x);

            denominador = deltaX * deltaX;
        }

        // Se for escolhido o tipo 2, entao sera calculado a derivada pelo metodo backward
        else if (tipo == 5){
            numerador = 
            ExecutarPython.converter_funcao(funcao, x)
            - 2 * ExecutarPython.converter_funcao(funcao, x - deltaX)
            + ExecutarPython.converter_funcao(funcao, x - 2 * deltaX);

            denominador = deltaX * deltaX;
        }

        // Se for escolhido o tipo 6, entao sera escolhido o metodo central
        else if (tipo == 6){
            numerador = ExecutarPython.converter_funcao(funcao, x + deltaX)
            - 2 * ExecutarPython.converter_funcao(funcao, x)
            + ExecutarPython.converter_funcao(funcao, x - deltaX);

            denominador = deltaX * deltaX;
        }

        // Calculando a reta secante
        double secante = numerador / denominador;
        return secante;

    }

    // Funcao para realizar a dedrivacao -> Aproximando o valor de deltaX para 0
    public double calcularDerivacao(double precisao, int tipo){

        // Criando as variaveis necessarias para a realizacao do calculo
        double resultadoAnterior = Float.POSITIVE_INFINITY;
        boolean loop = true;
        double secante = 0;

        // Loop para realizar diversos calculos aproximando deltaX de 0 ate chegar a uma diferenca minuscula
        while (loop){

            // Escolhendo a derivadacao
            if (tipo < 4){
                secante = derivadaPrimeira(tipo);
            }

            else{
                secante = derivadaSegunda(tipo);
            }

            // Se chegamos a um valor que seja menor que a precisao, entao chegamos a um valor favoravel
            if (Math.abs(resultadoAnterior - secante) < precisao){
                loop = false;
            }

            // Caso contrario, diminuimos deltaX
            this.setDeltaX(deltaX / 10);
            resultadoAnterior = secante;

        }

        // Quando chegamos a um resultado farovavel (proximo da derivada real) retornamos
        return secante;

    }
    
}
